File: superstitioMain/java/superstitio/cards/general/SkillCard/gainEnergy/TimeStop.java
package superstitio.cards.general.SkillCard.gainEnergy;

import com.evacipated.cardcrawl.mod.stslib.powers.interfaces.BetterOnApplyPowerPower;
import com.evacipated.cardcrawl.modthespire.lib.SpireInstrumentPatch;
import com.evacipated.cardcrawl.modthespire.lib.SpirePatch;
import com.evacipated.cardcrawl.modthespire.lib.SpirePrefixPatch;
import com.evacipated.cardcrawl.modthespire.lib.SpireReturn;
import com.megacrit.cardcrawl.actions.common.ApplyPowerAction;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import com.megacrit.cardcrawl.monsters.MonsterGroup;
import com.megacrit.cardcrawl.powers.AbstractPower;
import com.megacrit.cardcrawl.powers.NoDrawPower;
import javassist.CannotCompileException;
import javassist.expr.ExprEditor;
import javassist.expr.MethodCall;
import superstitio.DataManager;
import superstitio.SuperstitioImg;
import superstitio.cards.general.GeneralCard;
import superstitio.powers.AbstractSuperstitioPower;
import superstitio.powers.DelaySexualHeat;
import superstitio.powers.SexualHeat;

import java.util.Optional;

public class TimeStop extends GeneralCard {
    public static final String ID = DataManager.MakeTextID(TimeStop.class);

    public static final CardType CARD_TYPE = CardType.SKILL;

    public static final CardRarity CARD_RARITY = CardRarity.RARE;

    public static final CardTarget CARD_TARGET = CardTarget.SELF;

    private static final int COST = 2;
    private static final int MAGIC = 1;
    private static final int COST_UPGRADED_NEW = 1;


    public TimeStop() {
        super(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET);
        setupMagicNumber(MAGIC);
    }

    @Override
    public void use(AbstractPlayer player, AbstractMonster monster) {
        addToBot_applyPower(new TimeStopPower(AbstractDungeon.player, this.magicNumber));
        addToBot_applyPower(new NoDrawPower(player));
    }

    @Override
    public void upgradeAuto() {
        this.upgradeBaseCost(COST_UPGRADED_NEW);
    }

    @SuperstitioImg.NoNeedImg
    public static class TimeStopPower extends AbstractSuperstitioPower implements BetterOnApplyPowerPower {
        public static final String POWER_ID = DataManager.MakeTextID(TimeStopPower.class);
        public static final int sexualReturnRate = 2;

        public TimeStopPower(final AbstractCreature owner, int amount) {
            super(POWER_ID, owner, amount);
            this.loadRegion("time");
        }

        @Override
        public void updateDescriptionArgs() {
            setDescriptionArgs(amount, sexualReturnRate);
        }


        @Override
        public void atEndOfRound() {
            super.atEndOfRound();
            //if (!isPlayer) return;
            addToBot_AutoRemoveOne(this);
        }

        @Override
        public boolean betterOnApplyPower(AbstractPower power, AbstractCreature creature, AbstractCreature creature1) {
            if (power instanceof SexualHeat && power.amount > 0) {
                this.flash();
                this.addToBot(new ApplyPowerAction(this.owner, this.owner, new DelaySexualHeat(this.owner, power.amount * sexualReturnRate)));
                return false;
            }
            return true;
        }

        @SpirePatch(clz = AbstractCreature.class, method = "applyEndOfTurnTriggers")
        public static class TimeStopEndTurnPatch {
            @SpirePrefixPatch
            public static SpireReturn<Void> Prefix(AbstractCreature __instance) {
                Optional<TimeStopPower> timeStopPower =
                        __instance.powers.stream().filter(power -> power instanceof TimeStopPower).map(power -> (TimeStopPower) power).findAny();
                if (timeStopPower.isPresent()) {
                    AbstractPower noDraw = __instance.getPower(NoDrawPower.POWER_ID);
                    if (noDraw != null)
                        timeStopPower.get().addToBot_AutoRemoveOne(noDraw);
                    return SpireReturn.Return();
                }
                return SpireReturn.Continue();
            }
        }

        @SpirePatch(clz = MonsterGroup.class, method = "applyEndOfTurnPowers")
        public static class TimeStopEndRoundPatch {
            @SpireInstrumentPatch
            public static ExprEditor Instrument() {
                return new ExprEditor() {
                    public void edit(MethodCall m) throws CannotCompileException {
                        if (m.getClassName().equals(AbstractPower.class.getName()) && m.getMethodName().equals("atEndOfRound")) {
                            m.replace(String.format("if (!%s.shouldEscapeEndOfRound($0)){$_ = $proceed($$);}",
                                    TimeStopEndRoundPatch.class.getName()));
                        }

                    }
                };
            }

            public static boolean shouldEscapeEndOfRound(AbstractPower power) {
                if (power == null) return false;
                if (power instanceof TimeStopPower) return false;
                if (power.owner == null) return false;
                if (power.owner.powers == null) return false;
//                if (power instanceof NoDrawPower) return false;
                Optional<AbstractPower> timeStopPower =
                        power.owner.powers.stream().filter(power1 -> power1 instanceof TimeStopPower).findAny();
                return timeStopPower.isPresent();
            }
        }
    }
}

File: superstitioMain/java/superstitio/cards/general/SkillCard/gainEnergy/TimeStop.kt
package superstitio.cards.general.SkillCard.gainEnergy

import com.evacipated.cardcrawl.mod.stslib.powers.interfaces.BetterOnApplyPowerPower
import com.evacipated.cardcrawl.modthespire.lib.*
import com.megacrit.cardcrawl.actions.common.ApplyPowerAction
import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.monsters.AbstractMonster
import com.megacrit.cardcrawl.monsters.MonsterGroup
import com.megacrit.cardcrawl.powers.AbstractPower
import com.megacrit.cardcrawl.powers.NoDrawPower
import javassist.CannotCompileException
import javassist.expr.ExprEditor
import javassist.expr.MethodCall
import superstitio.DataManager
import superstitio.SuperstitioImg.NoNeedImg
import superstitio.cards.general.GeneralCard
import superstitio.cards.general.SkillCard.gainEnergy.TimeStop.TimeStopPower
import superstitio.powers.AbstractSuperstitioPower
import superstitio.powers.DelaySexualHeat
import superstitio.powers.SexualHeat
import superstitioapi.utils.addToBot_AutoRemoveOne
import superstitioapi.utils.setDescriptionArgs

class TimeStop : GeneralCard(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET)
{
    init
    {
        setupMagicNumber(MAGIC)
    }

    override fun use(player: AbstractPlayer?, monster: AbstractMonster?)
    {
        addToBot_applyPower(TimeStopPower(AbstractDungeon.player, this.magicNumber))
        addToBot_applyPower(NoDrawPower(player))
    }

    override fun upgradeAuto()
    {
        this.upgradeBaseCost(COST_UPGRADED_NEW)
    }

    @NoNeedImg
    class TimeStopPower(owner: AbstractCreature, amount: Int) : AbstractSuperstitioPower(POWER_ID, owner, amount),
        BetterOnApplyPowerPower
    {
        init
        {
            this.loadRegion("time")
        }

        override fun updateDescriptionArgs()
        {
            setDescriptionArgs(amount, sexualReturnRate)
        }


        override fun atEndOfRound()
        {
            super.atEndOfRound()
            //if (!isPlayer) return;
            addToBot_AutoRemoveOne(this)
        }

        override fun betterOnApplyPower(
            power: AbstractPower,
            creature: AbstractCreature,
            creature1: AbstractCreature
        ): Boolean
        {
            if (power is SexualHeat && power.amount > 0)
            {
                this.flash()
                this.addToBot(
                    ApplyPowerAction(
                        this.owner,
                        this.owner,
                        DelaySexualHeat(this.owner, power.amount * sexualReturnRate)
                    )
                )
                return false
            }
            return true
        }


        companion object
        {
            val POWER_ID: String = DataManager.MakeTextID(TimeStopPower::class.java)
            const val sexualReturnRate: Int = 2

            @SpirePatch2(clz = AbstractCreature::class, method = "applyEndOfTurnTriggers")
            object TimeStopEndTurnPatch
            {
                @SpirePrefixPatch
                @JvmStatic
                fun Prefix(__instance: AbstractCreature): SpireReturn<Void>
                {
                    val timeStopPower =
                        __instance.powers.stream().filter { power: AbstractPower? -> power is TimeStopPower }
                            .map { power: AbstractPower -> power as TimeStopPower }.findAny()
                    if (timeStopPower.isPresent)
                    {
                        val noDraw = __instance.getPower(NoDrawPower.POWER_ID)
                        if (noDraw != null) timeStopPower.get().addToBot_AutoRemoveOne(noDraw)
                        return SpireReturn.Return()
                    }
                    return SpireReturn.Continue()
                }
            }

            @SpirePatch2(clz = MonsterGroup::class, method = "applyEndOfTurnPowers")
            object TimeStopEndRoundPatch
            {
                @SpireInstrumentPatch
                @JvmStatic
                fun Instrument(): ExprEditor
                {
                    return object : ExprEditor()
                    {
                        @Throws(CannotCompileException::class)
                        override fun edit(m: MethodCall)
                        {
                            if (m.className == AbstractPower::class.qualifiedName && m.methodName == "atEndOfRound")
                            {
                                m.replace(
                                    String.format(
                                        "if (!%s.shouldEscapeEndOfRound($0)){\$_ = \$proceed($$);}",
                                        Companion::class.qualifiedName
                                    )
                                )
                            }
                        }
                    }
                }
            }

            @JvmStatic
            fun shouldEscapeEndOfRound(power: AbstractPower?): Boolean
            {
                if (power == null) return false
                if (power is TimeStopPower) return false
                if (power.owner == null) return false
                if (power.owner.powers == null) return false
                //                if (power instanceof NoDrawPower) return false;
                val timeStopPower =
                    power.owner.powers.stream().filter { power1: AbstractPower? -> power1 is TimeStopPower }
                        .findAny()
                return timeStopPower.isPresent
            }
        }
    }

    companion object
    {
        val ID: String = DataManager.MakeTextID(TimeStop::class.java)

        val CARD_TYPE: CardType = CardType.SKILL

        val CARD_RARITY: CardRarity = CardRarity.RARE

        val CARD_TARGET: CardTarget = CardTarget.SELF

        private const val COST = 2
        private const val MAGIC = 1
        private const val COST_UPGRADED_NEW = 1
    }
}
